test/functional/terminal/ex_terminal_spec.lua

```diff
@@ -51,10 +45,15 @@ describe(':terminal', function()
       ready $ echo 'hello' \ "world"                    |
                                                         |
       [Process exited 0]                                |
-                                                        |
-                                                        |
-                                                        |
       -- TERMINAL --                                    |
     ]])
   end)
+
+  it('ex_terminal() double-free #4554', function()
+    source([[
+      autocmd BufNew * set shell=foo
+      terminal]])
```

Comment:
This correctly invoked the ASan failure:

```
=================================================================
==12080==ERROR: AddressSanitizer: attempting double-free on 0x60600000c140 in thread T0:
    #0 0x4f1f72 in __interceptor_free (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f1f72)
    #1 0x2073d25 in xfree /home/travis/build/neovim/neovim/src/nvim/memory.c:105:3
    #2 0x1527fe5 in ex_terminal /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:9516:5
    #3 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #4 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #5 0x138a55d in do_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2611:3
    #6 0x1380347 in cmd_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2323:12
    #7 0x1380853 in ex_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2301:3
    #8 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #9 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #10 0x13dd51c in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:274:10
    #11 0x68414a in vim_command /home/travis/build/neovim/neovim/src/nvim/api/vim.c:43:3
    #12 0x599ed9 in handle_vim_command /home/travis/build/neovim/neovim/build/src/nvim/auto/msgpack_dispatch.c:2128:3
    #13 0x2348725 in on_request_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:481:19
    #14 0x118ff42 in queue_process_events /home/travis/build/neovim/neovim/src/nvim/event/queue.c:142:7
    #15 0x25cb500 in nv_event /home/travis/build/neovim/neovim/src/nvim/normal.c:7876:3
    #16 0x249de83 in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1144:12
    #17 0x3822e33 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:55:26
    #18 0x23e38b4 in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:464:3
    #19 0x1cfcbb7 in main /home/travis/build/neovim/neovim/src/nvim/main.c:538:3
    #20 0x2b01b582a76c in __libc_start_main /build/eglibc-rrybNj/eglibc-2.15/csu/libc-start.c:226
    #21 0x46b268 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x46b268)

0x60600000c140 is located 0 bytes inside of 54-byte region [0x60600000c140,0x60600000c176)
freed by thread T0 here:
    #0 0x4f1f72 in __interceptor_free (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f1f72)
    #1 0x2073d25 in xfree /home/travis/build/neovim/neovim/src/nvim/memory.c:105:3
    #2 0x27e3108 in free_string_option /home/travis/build/neovim/neovim/src/nvim/option.c:2165:5
    #3 0x2888a5f in did_set_string_option /home/travis/build/neovim/neovim/src/nvim/option.c:3142:7
    #4 0x2811cd2 in do_set /home/travis/build/neovim/neovim/src/nvim/option.c:1733:22
    #5 0x15120d4 in ex_set /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:9393:9
    #6 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #7 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #8 0x191d06f in apply_autocmds_group /home/travis/build/neovim/neovim/src/nvim/fileio.c:6711:5
    #9 0x18e7ff2 in apply_autocmds /home/travis/build/neovim/neovim/src/nvim/fileio.c:6358:10
    #10 0x70d323 in buflist_new /home/travis/build/neovim/neovim/src/nvim/buffer.c:1490:5
    #11 0x122cb59 in do_ecmd /home/travis/build/neovim/neovim/src/nvim/ex_cmds.c:2168:13
    #12 0x14a4f39 in do_exedit /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:6618:9
    #13 0x14b8769 in ex_edit /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:6541:3
    #14 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #15 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #16 0x13dd51c in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:274:10
    #17 0x1527eaf in ex_terminal /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:9513:3
    #18 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #19 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #20 0x138a55d in do_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2611:3
    #21 0x1380347 in cmd_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2323:12
    #22 0x1380853 in ex_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2301:3
    #23 0x141c2e0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #24 0x13ce683 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #25 0x13dd51c in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:274:10
    #26 0x68414a in vim_command /home/travis/build/neovim/neovim/src/nvim/api/vim.c:43:3
    #27 0x599ed9 in handle_vim_command /home/travis/build/neovim/neovim/build/src/nvim/auto/msgpack_dispatch.c:2128:3
    #28 0x2348725 in on_request_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:481:19
    #29 0x118ff42 in queue_process_events /home/travis/build/neovim/neovim/src/nvim/event/queue.c:142:7

previously allocated by thread T0 here:
    #0 0x4f2252 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f2252)
    #1 0x2072f35 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:59:15
    #2 0x2073981 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:93:15
    #3 0x20752b4 in xmallocz /home/travis/build/neovim/neovim/src/nvim/memory.c:168:15
    #4 0x20756a9 in xmemdupz /home/travis/build/neovim/neovim/src/nvim/memory.c:186:17
    #5 0x20798d8 in xstrdup /home/travis/build/neovim/neovim/src/nvim/memory.c:378:10
    #6 0x38237eb in vim_strsave /home/travis/build/neovim/neovim/src/nvim/strings.c:52:20
    #7 0x28b2a35 in set_string_option /home/travis/build/neovim/neovim/src/nvim/option.c:2334:7
    #8 0x27dbdd7 in set_option_value /home/travis/build/neovim/neovim/src/nvim/option.c:4653:14
    #9 0x6782dc in set_option_value_err /home/travis/build/neovim/neovim/src/nvim/api/private/helpers.c:817:25
    #10 0x653161 in set_option_value_for /home/travis/build/neovim/neovim/src/nvim/api/private/helpers.c:797:7
    #11 0x651c71 in set_option_to /home/travis/build/neovim/neovim/src/nvim/api/private/helpers.c:308:5
    #12 0x6933e8 in vim_set_option /home/travis/build/neovim/neovim/src/nvim/api/vim.c:394:3
    #13 0x5bec92 in handle_vim_set_option /home/travis/build/neovim/neovim/build/src/nvim/auto/msgpack_dispatch.c:2728:3
    #14 0x2348725 in on_request_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:481:19
    #15 0x118ff42 in queue_process_events /home/travis/build/neovim/neovim/src/nvim/event/queue.c:142:7
    #16 0x25cb500 in nv_event /home/travis/build/neovim/neovim/src/nvim/normal.c:7876:3
    #17 0x249de83 in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1144:12
    #18 0x3822e33 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:55:26
    #19 0x23e38b4 in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:464:3
    #20 0x1cfcbb7 in main /home/travis/build/neovim/neovim/src/nvim/main.c:538:3
    #21 0x2b01b582a76c in __libc_start_main /build/eglibc-rrybNj/eglibc-2.15/csu/libc-start.c:226

SUMMARY: AddressSanitizer: double-free ??:0 __interceptor_free
==12080==ABORTING
```

https://travis-ci.org/neovim/neovim/jobs/130438287